### PR TITLE
feat: viz migration script

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

There are some deprecate visualizations in Superset, for instance:
- nvd3 chart
- legacy pivot table
And these visualizations are often replaced with new versions: echart or pivot-table-v2

This script provides an approach to migrate legacy visualizations. 

currently, This script only migrates `pivot table` to `pivot table v2`

### Usage
```
$ superset migrate-viz -t pivot_table
```

### TESTING INSTRUCTIONS

provided unit test 

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/migrations/shared/migrate_viz.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    viz_type: str
+
+    def __init__(self, v1_data: str) -> None:
+        self.raw_data = v1_data
+        self.data = json.loads(v1_data)
+
+    def post_processing(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        raise NotImplementedError()
+
+    def migrate(self) -> str:
```
Comment (villebro) [inline superset/migrations/shared/migrate_viz.py:35]: It would probably be a good idea to support up and down migrations in this class

File: superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            idx += 1
+            print(f"Upgrading ({idx}/{total}): {slc.slice_name}#{slc.id}")
+            with open(
+                os.path.join(
+                    DATA_DIR, "migrate_viz_pivottable", f"{str(slc.uuid)}.json"
+                ),
+                "w",
+                encoding="utf-8",
+            ) as f:
+                f.write(slc.params)
```
Comment (villebro) [inline superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py:75]: Why is this needed, is it to make it possible to do a down migration? I don't think we need to write these to disk if the migrator would work in both directions.

**Review by villebro — COMMENTED** [review]
First pass comments. I think we should remove the need for persisting the original metadata

File: superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py (hunk 2 above)
Comment (zhaoyongjie) [reply superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py:75]: The main reason is that I can't find out which viz needs to do downgrade script.(The database might have a pivot table for both v1 and v2).

If all pivot table v2 do downgrade, it will cause all pivot tables (both v1 and v2) to execute downgrade.

Comment (villebro) [reply superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py:75]: Could we check for the presence of old properties, e.g. `groupby` (v1) vs `groupbyRows` (v2) and then only migrate the ones that are relevant?

Comment (zhaoyongjie) [reply superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py:75]: Let me leave an example
We have these charts in the database(before upgrade)
```
pivot_table_v1(a)
pivot_table_v1(b)
pivot_table_v2(c)
```

after upgrade
```
pivot_table_v2(a)
pivot_table_v2(b)
pivot_table_v2(c)
```

when we process downgrade script without dump file
```
pivot_table_v1(a)
pivot_table_v1(b)
pivot_table_v1(c)
```

the `pivot_table_v1(c)` unexpect to do downgrade.

Comment (villebro) [reply superset/migrations/versions/9ed569cf4ba9_migrate_pivottable_from_v1_to_v2.py:75]: Yes, I understand now 👍 I think skipping the downgrade in this case makes sense. Alternatively we could add a dummy field in the chart properties to indicate that it has been upgraded, after which it would be possible to downgrade the ones that had previously been upgraded. But I'd almost vote for just skipping the downgrade logic or accepting the fact that a downgrade will downgrade all v2 charts to v1.

**Comment by graceguo-supercat** [comment]
Hi @youngyjd and @villebro Thanks for the feature. I have a little concern before we merge this PR into master:
- Will this PR convert all `pivot table` to `pivot table v2`, and will not be able to roll back?
- How is feature parity for `pivot table` vs `pivot table v2`? see an issue reported here: https://apache-superset.slack.com/archives/G013HAE6Y0K/p1627595961063300
- could you let us (airbnb users) do more test before doing this db migration? 

Thanks!

**Comment by zhaoyongjie** [comment]
> Hi @youngyjd and @villebro Thanks for the feature. I have a little concern before we merge this PR into master:
> 
>     * Will this PR convert all `pivot table` to `pivot table v2`, and will not be able to roll back?
> 
>     * How is feature parity for `pivot table` vs `pivot table v2`? see an issue reported here: https://apache-superset.slack.com/archives/G013HAE6Y0K/p1627595961063300
> 
>     * could you let us (airbnb users) do more test before doing this db migration?
> 
> 
> Thanks!

I have discussed with @villebro, will create a `superset viz_migrate` command to do migration Instead of db upgrade to do this. I am working on this today.
